rm/rm.1
.\"
.\" Sccsid @(#)rm.1	1.14 (gritter) 1/24/05
.\" Parts taken from rm(1), Unix 7th edition:
.\" Copyright(C) Caldera International Inc. 2001-2002. All rights reserved.
.\" Copyright(C) Pindorama 2022-2024.
.\"
.\" SPDX-Licence-Identifier: Caldera
.TH RM 1 "3/19/24" "Heirloom Toolchest" "User Commands"
.SH NAME
rm \- remove directory entries
.SH SYNOPSIS
\fB/usr/5bin/rm\fR
[\fB\-df\&iRre\fR]
[\fIfile\fR ...]
.br
\fB/usr/5bin/s42/rm\fR
[\fB\-f\&iRr\fR]
\fIfile\fR ...
.SH DESCRIPTION
Rm removes the entries
for one or more files from a directory.
If an entry was the last link to the file,
the file is destroyed.
Removal of a file requires write permission in its directory,
but neither read nor write permission on the file itself.
.PP
If a file has no write permission
and the standard input is a terminal,
its permissions are printed
and a line is read from the standard input.
If that line begins with `y'
the file is deleted,
otherwise the file remains.
.PP
When
.B rm
encounters a symbolic link,
the link will be removed,
but its target will remain in the file hierarchy.
.PP
The
.B rm
command accepts the following options:
.TP
.B \-d
Attempts to delete empty directories and
other types of files.
This option has been introduced by 4.4BSD
in 1993 and is an extension to the
standard.
This option is only present at
.BR /usr/5bin/rm .
.TP
.B \-e
Displays a message after deleting each file.
This option has been introduced by IBM AIX(R)
and is an extension to the standard (see NOTES).
This option is only present at
.BR /usr/5bin/rm .
.TP
.B \-f
No questions are asked
and neither error messages are printed
nor the exit status is affected
in case of nonexistent files.
.B /usr/5bin/rm
and
.B /usr/5bin/s42/rm
will not print an error message
if removal of a file failed.
.br
For
.B /usr/5bin/rm
specifically,
an error message will not printed
if there's not a file being informed
as a parameter. This payload has been
introduced by POSIX.1-2008.
.br
For
.B /usr/5bin/posix/rm
and
.BR /usr/5bin/posix2001/rm ,
any previous occurrences of the
.B \-i
option are ignored.
.TP
.B \-i
Causes
.B rm
to ask for confirmation
before deleting any file,
and,
unter
.BR \-r ,
whether to examine each directory.
For
.B /usr/5bin/posix/rm
and
.BR /usr/5bin/posix2001/rm ,
any previous occurrences of the
.B \-f
option are ignored.
.TP
.B \-R
Same as
.BR \-r .
This option has been introduced by POSIX.2.
.TP
.B \-r
.B Rm
will recursively delete the entire contents
of directories given as operands,
and the directory itself.
Without this option,
an error comment is printed
if a designated file is a directory.
.SH "ENVIRONMENT VARIABLES"
.TP
.B SYSV3
Causes the text of some messages to be changed.
.SH "SEE ALSO"
rmdir(1),
unlink(2),
rmdir(2)
.SH NOTES
It is forbidden to remove the files `.' and `..'
merely to avoid the antisocial consequences
of inadvertently doing something like `rm \-r .*'.
.PP
With
.B /usr/5bin/rm
and
.BR /usr/5bin/s42/rm ,
a single `\-' can be used to indicate the end of the options list,
as with `rm \- \-file'.
If `\-\-' is used to terminate the options list, though,
a following `\-' will be interpreted as a file name,
so `rm \-\- \-' will remove the file `\-'.
.PP
Option \-e is an extension and should not be
cluelessly used on shell scripts. It apparently
appeared first (and only) on IBM AIX(R) UNIX, but
it works pretty much like as BSD/GNU's \-v option.
.br
If verbosing is desired, a measure to abide with
would be to use a boilerplate function that deals
with different
.B rm(1)
implementations.
An example:
.RS 2
.sp
.nf
_rmv() {
    if test \`getconf HEIRLOOM_TOOLCHEST_VERSION\` \-gt 20240220 \-o "x\`uname -s\`" \\= \'xAIX\'; then
        rm \-re "$@"
    elif (rm \-\-help 2>&1 | egrep \'\\[\\-.*v.*\\]|GNU\' 2>&1 > /dev/null); then
        rm \-rv "$@"
    else
        if test \-n "$1"; then
            for file do
                if test \-L "$file" \-o \-f "$file"; then
                    rm "$file" \\
                    && printf \'rm: removed file %s\\n\' "$file"
                elif test \-d "$file"; then
                    find "$file" \-type f \-print \\
                    | (while read sfile; do
                        rm "$sfile" \\
                        && printf \'rm: removed file %s\\n\' \\
                            "$sfile"
                      done) \\
                    && find "$file" \-type d \-print \\
                    | (while read directory; do
                        rmdir "$directory" \\
                        && printf \'rm: removed directory %s\\n\' \\
                            "$directory"
                    done)
                fi
            done
        fi
    fi
}
.fi
.sp
.RE
Above there is a shell function that
works per checking which implementation
is being used and then use the correct
option for it. In case of the
.B rm(1)
program being strictly POSIX, it will
emulate \-e/\-v payloads per using
.B find(1)
along with the default
.B rm(1)
and
.B rmdir(1)
commmands inside a loop. See
.B sh(1)
for more information about the
shell language structure itself.
It can be said that the strict POSIX
fallback is fairly slower than the other
options, but it works even in the
crudest UNIX-compatible environments.
.br
This is just an example that can be
freely used and improved; I will not
delve much into this merit.


CHANGES
Release 250209
* cpio: Fix manual page formatting.
* apply: Fixed the command string magic character computation algorithm that
  got garbled after a refactor made in the name of performance.
* build/mk.config, chroot, nawk: Adding -ltinfo for static compilation with
  glibc and making changes to adhere to C standards (patch by Gabriel Souza).
* [NEW] wall, write: New implementation ported from UNIX 7th (a.k.a. 32V),
  bulletproofed against buffer overflows and the CVE-2024-28085, which is
  available as the 'ssafe()' function at libcommon.
* rm: Alias '-e' flag to more known '-v' for verbose.
* date, shl: Fix compilation on platforms where (struct utmpx).ut_tv is not
  of type 'timeval'.
* rev: UTF-8 support, borrowed from suckless sbase. In varietate concordia.
* [NEW] apply, pick: Complete implementation for both apply and pick, from
  Research UNIX 8th and 10th respectively; extended payloads for pick in
  comparison with the original implementation.
* seq: Support for counting from negative numbers to positive or making a
  negative sequence.
* rev: Full refactor (made by Samuel Brederodes).
* Fixes for building dynamically on Arch Linux and other GNU C
  Library-based Linux systems which ships exclusively dynamic versions
  of its libraries, also remove `readlink_ucb' object file at `readlink/'
  build directory when executing `make clean' (patch by Molly A. McCollum).
* Remove `register's from legacy code that was living unchanged since at
  least the 80's, since modern linters started complaining about it and
  it is not used as much on most common architectures such as x86
  (patch by Samuel Brederodes).
* watch: `use_default_colors()', do not force black and white.
* [NEW] rev: New implementation, not capable of UTF-8 yet.
* cpio, diff, tabs: Fix duplicated definitions (made/reported by crumbtoo,
  merged by me, Luiz Antônio Rangel). Thanks! --- or better: Merci!
* rm: Added real support for the POSIX 1003.1(2008) standard, implemented
  IBM AIX(R)-style verbosing and 4.4BSD's '-d' option for directories ---
  rmdir(1)-esque payload --- as extensions to the standard.
* seq: Made its payload similar to UNIX v8's implementation, but with some
  extensions such as a fast algorithm for -w or -s --- which was already
  present before the (almost) full rewrite.
* timeout: avoid buffer overrun, also did some changes at the source code,
  such as cleaning up comments and initializing 'commandv' and 'fst_commandv'.
* watch: Used esr's deheader for removing unnecessary headers, also use
  pfmt(stderr, MM_ERROR, ...)  instead of prerror().
* watch, timeout: Some small corrections at the manual page.
* cpio: Skip over zip extra headers with fewer than four bytes (the minimum
  length for them to contain any information) instead of reporting a bad
  header and terminating.
* pg: Handle cases where standard output is a terminal but not open for
  reading (bug reported by Keith Doyle).
* cpio, tar: If an archive contains a file that is a hardlink to itself on
  extraction, do nothing at all instead of discarding its data. This mainly
  happens if a file with more than one link is included under the same name
  in an archive twice. Since special situations on extraction such as bind
  mounts on Linux may lead to this scenario even in cases where both link
  names are different, this is now treated like a successful link() step
  (bugreport by Thomas Graf).
* Made the getopt() substitute work with the getopt$UNIX2003() function
  redirection mechanism on Mac OS X Leopard.
* mknod, pgrep, ps: Mac OS X Leopard compilation fixes.
* w, whodo: On a 32-bit system where /proc/kcore is larger than 2 GB,
  avoid a harmless error message (bug reported by Pierre Gentile).
* ptime: Works on recent Linux kernels again.
* col: The behavior of the -x flag has been inverted to actually match
  the documentation, i.e. to suppress conversion of spaces to tabs (patch
  by Markus Schnalke).
* cut: With "-s", print the last field of a line that was also the last
  field specified with "-f" (bugreport by Pierre Gentile).

Release 240220
* [NEW] seq: Stone-simple implementation; needs some improvement yet.
* [NEW] watch: Complete, simple and stable implementation.
* [NEW] timeout: Complete implementation with extended payloads.
* readlink: Different help messages for both UCB and default variants.

Release 230503
* chroot: "docs: Update date at chroot.8 manual page."
* chroot: "chore: Remove comment, will move to GitHub"
* chroot: "docs: Rewrote chroot(8) manual page"
* chroot: "chore: Add ``usage()`` function at chroot.c"
* tabs: "fix: [t,g]roff formatting at tabs/fspec.5."
* readlink: "docs: Created new manual page for the non-UCB variant. (#24)"
* Makefile.mk: "fix: Don't symbolic/hard link readlink from UCBDIR to DEFDIR anymore"
* readlink: "docs: Add comments clarifying the C preprocessor directives."
* readlink: "fix: Ignore ``-n`` option when operating multiple files."
* readlink: "feat: Create non-UCB/GNU-alike variant."
* readlink: "fix: Made ``name_size`` private again, so ``pfmt()`` won't go segmentation fault."
* readlink: "feat: First step, start modularizing the code."
* libcommon: "fix: getdirentries replacement for OBSD >5.4"
* fix-me: For new releases, finish porting for OpenBSD and implement watch(1) and timeout(1).

Release 230305
* getconf: "Now _NPROCESSORS_ONLN and _NPROCESSORS_CONF are implemented."
* (*): "Removed the remaining licence headers, now every licence is referenced by its SPDX identifier."

Release 230228
* [NEW] readlink: New implementation, extensively based on OpenBSD's.
* readlink: It's installed at /usr/ucb per default and symbolic/hard linked to /usr/5bin.

Release 230120
* Notes for the library directories if linking against the GNU C library.
* rm: Consistency within standards: the SUS program shouldn't have the POSIX 1003.1(2008) fix, just the main program.
* Made $(ROOT) not hardcoded anymore.

rm/rm.c
/*
 * rm - remove directory entries
 *
 * Gunnar Ritter, Freiburg i. Br., Germany, July 2002.
 * Luiz Antônio Rangel, Brazil, March 2024.
 */
/*
 * Copyright (c) 2003 Gunnar Ritter
 * Copyright (c) 2022-2024: Luiz Antônio Rangel (takusuman)
 *
 * SPDX-Licence-Identifier: Zlib
 */

#if __GNUC__ >= 3 && __GNUC_MINOR__ >= 4 || __GNUC__ >= 4
#define	USED	__attribute__ ((used))
#elif defined __GNUC__
#define	USED	__attribute__ ((unused))
#else
#define	USED
#endif
#ifdef	SUS
static const char sccsid[] USED = "@(#)rm_sus.sl	2.23 (gritter) 11/17/05";
#else
static const char sccsid[] USED = "@(#)rm.sl	2.23 (gritter) 11/17/05";
#endif

#include	<sys/types.h>
#include	<sys/stat.h>
#include	<sys/wait.h>
#include	<sys/resource.h>
#include	<fcntl.h>
#include	<unistd.h>
#include	<stdio.h>
#include	<string.h>
#include	<stdlib.h>
#include	<errno.h>
#include	<libgen.h>
#include	<dirent.h>
#include	<limits.h>
#include	<signal.h>
#include	<stdarg.h>

#include	"getdir.h"

struct	level {
	struct getdb	*l_db;
	int	l_fd;
};

static struct level	*levels;	/* visited directories */
static size_t	maxlevel;		/* maximum level visited */
static unsigned	errcnt;			/* count of errors */
static int	ontty;			/* stdin is a tty */
static int	fflag;			/* force */
static int	iflag;			/* ask for confirmation */
static int	rflag;			/* recursive */
#if !defined(SUS) && !defined(SYSV3)
static int	eflag;			/* Displays a message after deleting each file. */
static int	dflag;			/* Attempt to remove empty directories. */
#endif
static char	*progname;		/* argv[0] to main() */
static char	*path;			/* full path to current file */
static size_t	pathsz;			/* allocated size of path */
extern int	sysv3;			/* emulate SYSV3 behavior */

static int	subproc(size_t, const char *, int);

static void *
srealloc(void *vp, size_t nbytes)
{
	void *p;

	if ((p = realloc(vp, nbytes)) == NULL) {
		write(2, progname, strlen(progname));
		write(2, ": Insufficient memory.\n", 23);
		exit(077);
	}
	return p;
}

static void *
smalloc(size_t nbytes)
{
	return srealloc(NULL, nbytes);
}

static void
usage(void)
{
	fprintf(stderr,
#if defined(SUS) || defined(SYSV3)
		"usage: %s [-fir] file ...\n",
#else
		"usage: %s [-dfire] file ...\n",
#endif
		progname);
	exit(2);
}

static void
msg(const char *fmt, ...)
{
	va_list	ap;

	if (!sysv3)
		fprintf(stderr, "%s: ", progname);
	va_start(ap, fmt);
	vfprintf(stderr, fmt, ap);
	va_end(ap);
}

static int
confirm(void)
{
	int yes = 0;
	char c;

	if (read(0, &c, 1) == 1) {
		yes = (c == 'y' || c == 'Y');
		while (c != '\n' && read(0, &c, 1) == 1);
	}
	return yes;
}

static void
setlevel(int level, struct getdb *db, int fd)
{
	if (level >= maxlevel)
		levels = srealloc(levels, (maxlevel=level+16) * sizeof *levels);
	levels[level].l_db = db;
	levels[level].l_fd = fd;
}

static size_t
catpath(size_t pend, const char *base)
{
	size_t	blen = strlen(base);

	if (pend + blen + 2 >= pathsz)
		path = srealloc(path, pathsz = pend + blen + 16);
	if (pend == 0 || path[pend-1] != '/')
		path[pend++] = '/';
	strcpy(&path[pend], base);
	return pend + blen;
}

static int
rmfile(const char *base, const struct stat *sp)
{
	if (iflag) {
		msg(sysv3 ? "%s ? " : "remove  %s: (y/n)? ", path);
		if (confirm() == 0)
			return 0;
	} else if ((sp->st_mode&S_IFMT) != S_IFDIR &&
			(sp->st_mode&S_IFMT) != S_IFLNK &&
			ontty && !fflag &&
			access(base, W_OK) < 0) {
		msg("%s: %o mode ? ", path, (int)(sp->st_mode & 0777));
		if (confirm() == 0)
			return 0;
	}
	if (((sp->st_mode&S_IFMT) == S_IFDIR ? rmdir : unlink)(base) < 0) {
#ifndef	SUS
		if ((sp->st_mode&S_IFMT) == S_IFDIR || !fflag || iflag)
#endif
			fprintf(stderr, (sp->st_mode&S_IFMT) == S_IFDIR ?
				"%s: Unable to remove directory %s\n%s\n" :
				"%s: %s not removed.\n%s\n",
				progname, path, strerror(errno));
		errcnt |= 1;
		return -1;
	}
#if !defined(SUS) && !defined(SYSV3)
	if (eflag) {
		msg(((sp->st_mode&S_IFMT) == S_IFDIR
			? "removing directory %s\n"
			: "removing %s\n"),
				path);
	}
#endif
	return 0;
}

static void
rm(size_t pend, const char *base, const int olddir, int ssub, int level)
{
	struct stat st;

	if (lstat(base, &st) < 0) {
		if (fflag == 0 || errno != ENOENT) {
			if (sysv3)
				fprintf(stderr, "%s: %s non-existent\n",
						progname, path);
			else
				fprintf(stderr, "%s: %s\n",
						path, strerror(errno));
			errcnt |= 4;
		}
		return;
	}
	if ((st.st_mode&S_IFMT) == S_IFDIR) {
		if (rflag) {
			struct direc *dp;
			struct getdb *db;
			int df, err;

			if (ssub == 0 && (iflag
#ifdef	SUS
					|| (!fflag && ontty &&
						access(base, W_OK) < 0)
#endif
					)) {
				msg("directory %s: ? ", path);
				if (confirm() == 0)
					return;
			}
			if ((df = open(base,
							O_RDONLY
#ifdef	O_DIRECTORY
							| O_DIRECTORY
#endif
#ifdef	O_NOFOLLOW
							| O_NOFOLLOW
#endif
						)) < 0 ||
					(db = getdb_alloc(base, df)) == NULL) {
				if (errno == EMFILE) {
					int	sres;

					sres = subproc(pend, base, level);
					if (sres >= 0) {
						errcnt |= sres;
						goto remove;
					}
				}
				/*
				 * Maybe the directory is empty and can just
				 * be removed.
				 */
				if (rmfile(base, &st) < 0) {
					fprintf(stderr,
						"%s: cannot read "
						"directory %s\n",
						progname, path);
					errcnt |= 4;
				}
				return;
			}
			if (fchdir(df) < 0) {
				if (rmfile(base, &st) < 0) {
					fprintf(stderr,
						"%s: cannot chdir to %s\n",
							progname, path);
					errcnt |= 4;
				}
				getdb_free(db);
				close(df);
				return;
			}
			setlevel(level, db, df);
			while ((dp = getdir(db, &err)) != NULL) {
				if (dp->d_name[0] == '.' &&
						(dp->d_name[1] == '\0' ||
						 dp->d_name[1] == '.' &&
						 dp->d_name[2] == '\0'))
					continue;
				rm(catpath(pend, dp->d_name), dp->d_name,
						df, 0, level + 1);
				path[pend] = '\0';
			}
			if (err) {
				fprintf(stderr,
					"%s: error reading directory %s\n",
					progname, path);
				errcnt |= 4;
			}
			if (olddir >= 0 && fchdir(olddir) < 0) {
				fprintf(stderr, "%s: cannot change backwards\n",
						progname);
				exit(1);
			}
			getdb_free(db);
			close(df);
#if !defined(SUS) && !defined(SYSV3)
		/* Latter priority than rflag. */
		} else if (dflag) {
			if (rmfile(base, &st) < 0) {
				/*
				 * rmfile() will print a error
				 * message per si if needed.
				 */
				errcnt |= 1;
			}
			return;
#endif
		} else {
			fprintf(stderr, "%s: %s directory\n", progname, path);
			errcnt |= 1;
			return;
		}
	}
	if (ssub == 0)
	remove:	rmfile(base, &st);
}

/*
 * POSIX.2 specifies that rm shall not fail because it exceeds the
 * maximum number of open file descriptors. It is not possible to
 * use path names and chdir("..") instead since the tree might be
 * moved during operation and we'd then start to remove files outside.
 * Creating a new process makes it possible to circumvent the OPEN_MAX
 * limit in a safe way.
 */
static int
subproc(size_t pend, const char *base, int level)
{
	pid_t	pid;

	switch (pid = fork()) {
	case 0: {
		int	i;

		errcnt = 0;
		for (i = 0; i < level - 1; i++) {
			getdb_free(levels[i].l_db);
			close(levels[i].l_fd);
		}
		rm(pend, base, -1, 1, 0);
		_exit(errcnt);
	}
	default: {
		int status;

		while (waitpid(pid, &status, 0) != pid);
		if (status && WIFSIGNALED(status)) {
			/*
			 * If the signal was sent due to a tty keypress,
			 * we should be terminated automatically and
			 * never reach this point. Otherwise, we terminate
			 * with the same signal, but make sure that we do
			 * not overwrite a possibly generated core file.
			 * This results in nearly the usual behavior except
			 * that the shell never prints a 'core dumped'
			 * message.
			 */
			struct rlimit	rl;

			rl.rlim_cur = rl.rlim_max = 0;
			setrlimit(RLIMIT_CORE, &rl);
			raise(WTERMSIG(status));
			pause();
		}
		return status ? WEXITSTATUS(status) : 0;
	}
	case -1:
		return -1;
	}
}

static char *
dotdot(const char *s)
{
	const char	*sp;

	for (sp = s; *sp; sp++);
	sp--;
	while (sp > s && *sp == '/')
		sp--;
	if (sp[0] == '.' && (sp == s || (sp > s && sp[-1] == '/')))
		return ".";
	if (&sp[-1] >= s && sp[0] == '.' && sp[-1] == '.' && (&sp[-1] == s ||
				&sp[-1] > s && sp[-2] == '/'))
		return "..";
	return NULL;
}

int
main(int argc, char **argv)
{
	int i, startfd = -1, illegal = 0;
	char *options = "";

#ifdef	__GLIBC__
	putenv("POSIXLY_CORRECT=1");
#endif
	progname = basename(argv[0]);
	if (getenv("SYSV3") != NULL)
		sysv3 = 1;

#if defined(SUS) || defined(SYSV3)
	options = "fiRr";
#else
	options = "dfiRrev";
#endif

	while ((i = getopt(argc, argv, options)) != EOF) {
		switch (i) {
		case 'f':
			fflag = 1;
#ifdef	SUS
			iflag = 0;
#endif
			break;
		case 'i':
#ifdef	SUS
			fflag = 0;
#endif
			iflag = 1;
			break;
		case 'R':
		case 'r':
			rflag = 1;
			break;
#if !defined(SUS) && !defined(SYSV3)
		case 'd':
			dflag = 1;
			break;
		case 'e':
		case 'v':
			eflag = 1;
			break;
#endif
		default:
			illegal = 1;
		}
	}
	if (illegal)
		usage();
#ifndef	SUS
	if (argv[optind] && argv[optind][0] == '-' && argv[optind][1] == '\0' &&
			(argv[optind-1][0] != '-' || argv[optind-1][1] != '-' ||
			 argv[optind-1][2] != '\0'))
		optind++;
#endif

#if defined(SUS) || defined(SYSV3)
	if (optind >= argc)
#else
	if (optind >= argc && (!fflag || iflag)) /* Ha, touché. */
#endif
		usage();
	ontty = isatty(0);
	if (rflag && (startfd = open(".", O_RDONLY)) < 0) {
		fprintf(stderr, "%s: cannot open current directory\n",
				progname);
		exit(1);
	}
	while (optind < argc) {
		if (dotdot(argv[optind]) != NULL) {
			if (fflag == 0)
				fprintf(stderr, "%s of %s is not allowed\n",
					progname, argv[optind]);
			errcnt |= 1;
		} else {
			if (path)
				free(path);
			i = strlen(argv[optind]);
			strcpy(path = smalloc(pathsz = i + 1), argv[optind]);
			rm(i, argv[optind], startfd, 0, 0);
		}
		optind++;
	}
	return errcnt;
}
